python/tvm/script/parser.py

```diff
@@ -440,8 +463,23 @@ def check_decorator(decorators: List[ast.Expr]) -> bool:
         # add parameters of function
         for arg in node.params:
             arg_var = tvm.te.var(arg.name, self.parse_type(arg.ty, arg))
-            self.context.update_symbol(arg.name, arg_var, node)
             self.context.func_params.append(arg_var)
+            # Note that this case is for T.match_buffer syntax sugar
+            if isinstance(arg.ty, (ast.TypeCall, ast.TypeApply)):
+                result = self.transform(arg.ty)
+                if not isinstance(result, buffer.Buffer):
+                    self.report_error(
+                        "The result type of evaluating TypeCall and TypeApply stmt"
+                        f" is wrong: {type(result)}. It should be a Buffer",
+                        node.span,
+                    )
+                arg_name_with_handle = arg.name + "_handle"
+                new_arg_var = tvm.te.var(arg_name_with_handle, self.parse_type(arg.ty, arg))
```

Comment:
`self.transform(arg.ty)` and `parse_type` is kind of duplication, dtype of `te.var` here is actually the same as `buffer.dtype`